### ISSUE TITLE
[Bug] Throw exception though transaction committed successfully.

### ISSUE THREAD

**Issue by thetumbled** [apache/pulsar#19629]
### Search before asking

- [X] I searched in the [issues](https://github.com/apache/pulsar/issues) and found nothing similar.

### Version

2.9

### Minimal reproduce step

transactional produce 100 million messages, and restart brokers every 120s for 50 times.

### What did you expect to see?

we expect that there is no any message duplication.

### What did you see instead?

there exists message duplication.

### Anything else?

i find some reasons accounting for this problem.
```
2023-02-23T15:52:00,379+0800 [pulsar-client-internal-5-1] ERROR org.apache.pulsar.testclient.PerformanceProducer - Commit transaction failed with exception :
java.util.concurrent.CompletionException: org.apache.pulsar.client.api.transaction.TransactionCoordinatorClientException$InvalidTxnStatusException: Expect Txn `(7,24726)` to be in COMMITTING status but it is in COMMITTED status
```
when producer commit transaction to broker, the broker throw exceptions though the transaction have been committed successfully. 
Then the producer receive the exception and thought that it fail to commit the transaction, as a result producer resend the messages in the transaction, which account for the message duplication.

<img width="645" alt="image" src="https://user-images.githubusercontent.com/52550727/221130309-d34ed94f-173b-459a-a3f5-194bd74441cb.png">
<img width="640" alt="image" src="https://user-images.githubusercontent.com/52550727/221130477-2e0c7abd-ec70-4771-98cd-e4e66a9532f8.png">
<img width="915" alt="image" src="https://user-images.githubusercontent.com/52550727/221130554-521a3586-0d49-49b1-af1f-2c0bee721fb6.png">

### Are you willing to submit a PR?

- [X] I'm willing to submit a PR!

**Comment by congbobo184** [comment]
I don't quite understand the impact of repeated commits?

**Comment by nicoloboschi** [comment]
>when producer commit transaction to broker, the broker throw exceptions though the transaction have been committed successfully.

@thetumbled 
why the commit happens twice? this shouldn't happen in a normal situation

**Comment by thetumbled** [comment]
> I don't quite understand the impact of repeated commits? Does it duplicate the message?

yes, repeated commits will result into exceptions. As client receive exceptions when committing, it will resend messages in a new transaction because it think that the old transaction is committed failed.

**Comment by thetumbled** [comment]
> > when producer commit transaction to broker, the broker throw exceptions though the transaction have been committed successfully.
> 
> @thetumbled why the commit happens twice? this shouldn't happen in a normal situation

Restarting the broker may cause multiple commits. Detailed reason refer to last paragraph.
> Finally, the analysis above do not explain why there are two commit request arrive to broker. I will explain this following.

**Comment by thetumbled** [comment]
> > > I don't quite understand the impact of repeated commits? Does it duplicate the message?
> > 
> > 
> > yes, repeated commits will result into exceptions. As client receive exceptions when committing, it will resend messages in a new transaction because it think that the old transaction is committed failed.
> 
> I am not sure it's a good use case.
> 
> let's see this situation :
> 
> 1. consumer one message
> 2. produce one message with transactionOne
> 3. consumer ack with transactionOne
> 4. commit this transaction
> 5. tc handle this commit request and then change the status to committed
> 6. tc unload, and don't send the response to client
> 
> how does the client do at this time? produce again?
> 
> your case is used for local transaction link (sql transaction) ? same as rocketMq?

if TC do not send the response to client, client should retry to commit the transaction. if client can't know whether the transaction is committed successfully, how can we ensure exactly-once semantics? 
I have redeveloped the `PerformanceProducer` to support for checking whether the exactly-once semantics is guaranteed. And in my experiment, unloading topics will not result into message duplication, but restarting broker will. I can share my experiment results if you want.

**Comment by congbobo184** [comment]
like the above scene
7. tc reload, the TransactionOne has been committed, so the transactionOne can't be in TC
8. client recommitted this transaction will throw TransactionNotFoundException
9. how user know the transactionOne committed or timeout aborted?

Currently, Pulsar transaction does not support binding with local transactions, the server needs to notify the client, if you want to binding local transaction.

current we only support
1. consumer ack with transaction
2. producer produce with transaction
3. end this transaction

if this transaction failed, the consumer will rereceive this message and process it again. so in this case, we don't need to let the user can know if the transaction was committed or aborted, only redeliver this message again is enough.

**Comment by thetumbled** [comment]
> like the above scene 7. tc reload, the TransactionOne has been committed, so the transactionOne can't be in TC 8. client recommitted this transaction will throw TransactionNotFoundException 9. how user know the transactionOne committed or timeout aborted?
> 
> Currently, Pulsar transaction does not support binding with local transactions, the server needs to notify the client, if you want to binding local transaction.
> 
> current we only support
> 
> 1. consumer ack with transaction
> 2. producer produce with transaction
> 3. end this transaction
> 
> if this transaction failed, the consumer will rereceive this message and process it again. so in this case, we don't need to let the user can know if the transaction was committed or aborted, only redeliver this message again is enough.

If transactionOne is committed successfully and client reproduce messages in transactionOne, message duplication will occur, which violate the exactly-once semantics.
And i have mention `TransactionNotFoundException` in the analysis in the issue, which can not be fixed by https://github.com/apache/pulsar/pull/19662. But I have figure out a simple patch. We can hold txnMeta in txnMetaMap for transaction-timeout. If the transaction time out, client can not issue the commit request anymore, so we do not need to hold the txnMeta in memory anymore. Currently, the txnMeta will be removed instantly after the transaction is terminated, which will result into `TransactionNotFoundException`.

**Comment by congbobo184** [comment]
if the transaction timeout, we store it in the memory, what time the transaction meta can be removed from the memory?

**Comment by thetumbled** [comment]
> if the transaction timeout, we store it in the memory, what time the transaction meta can be removed from the memory?
if TC unload, how does we know this transaction aborted by client or by timeout?

I means that we just hold the txnMetaData for transaction-timeout, a fixed client-configured duration, for example 5s. 

> why enable deduplication can't ensure exactly-once semantics?

see https://streamnative.io/blog/exactly-once-semantics-transactions-pulsar
```
While powerful, Idempotent producer only solves a narrow scope of challenges for exactly-once semantics. There are still many other challenges it doesn’t resolve. For example, there is no atomicity when a producer attempts to produce messages to multiple topics. 
```

**Comment by congbobo184** [comment]
> I means that we just hold the txnMetaData for transaction-timeout, a fixed client-configured duration, for example 5s.
I don't quite understand, does the client need to configure two timeouts?

can you show me your use example? As far as I guess, your example doesn't contain any consumers?

If you want to ensure multi-topic produce exactly-once semantics, now pulsar transaction can't ensure.

**Comment by thetumbled** [comment]
> I don't quite understand, does the client need to configure two timeouts?
I have implement this in PR https://github.com/apache/pulsar/pull/19662. 

> can you show me your use example? As far as I guess, your example doesn't contain any consumers?
> If you want to ensure multi-topic produce exactly-once semantics without consumer, now pulsar transaction can't ensure.

yes we do not contain consumers in our experiment. We want to use pulsar transaction to ensure producer-side exactly-once semantics, and use flink-connector to provide consumer-side exactly-once semantics. 
Later i will  issue a PR for supporting checkout whether transactional producer ensure exactly-once semantics.

**Comment by congbobo184** [comment]
> > I don't quite understand, does the client need to configure two timeouts?
> 
> I have implemented this in PR #19662.
> 
> > can you show me your use example? As far as I guess, your example doesn't contain any consumers?
> > If you want to ensure multi-topic produce exactly-once semantics without consumer, now pulsar transaction can't ensure.
> 
> yes we do not contain consumers in our experiment. We want to use pulsar transaction to ensure producer-side exactly-once semantics, and use flink-connector to provide consumer-side exactly-once semantics. Later i will issue a PR for supporting checkout whether transactional producer ensure exactly-once semantics.

I understand what you mean, but this may be a new feature. #19662 fix some situations but can resolve all problems. we may need retention the ended transaction and provide a search API to check the txn status. It is a new feature, are you interested in doing it?

**Comment by thetumbled** [comment]
sure. I will try to implement this feature.
